osbenchmark/workload_generator/corpus.py

```diff
@@ -64,49 +67,187 @@ def extract(client, output_path, index, number_of_docs_requested=None):
+            "[%d] total docs in index [%s]. Extracting [%s] docs.",
+            number_of_docs,
+            index,
+            total_docs,
+        )
         docs_path = get_doc_outpath(output_path, index)
-        dump_documents(client, index, get_doc_outpath(output_path, index, "-1k"), min(total_docs, 1000), " for test mode")
-        dump_documents(client, index, docs_path, total_docs)
+        dump_documents(
+            concurrent,
+            client,
+            index,
+            get_doc_outpath(output_path, index, "-1k"),
+            min(total_docs, 1000),
+            "for test mode",
+        )
+        dump_documents(concurrent, client, index, docs_path, total_docs)
         return template_vars(index, docs_path, total_docs)
     else:
-        logger.info("Skipping corpus extraction fo index [%s] as it contains no documents.", index)
+        logger.info(
+            "Skipping corpus extraction fo index [%s] as it contains no documents.",
+            index,
+        )
         return None
 
 
-def dump_documents(client, index, out_path, number_of_docs, progress_message_suffix=""):
-    # pylint: disable=import-outside-toplevel
-    from opensearchpy import helpers
+def dump_documents_range(
+    pbar,
+    client,
+    index,
+    out_path,
+    start_doc,
+    end_doc,
+    total_docs,
+    progress_message_suffix="",
+):
+    """
+    Extract documents in the range of start_doc and end_doc and write to individual files
+
+    :param client: OpenSearch client used to extract data
+    :param index: Name of OpenSearch index to extract documents from
+    :param out_path: Destination directory for corpus dump
+    :param start_doc: Start index of the document chunk
+    :param end_doc: End index of the document chunk
+    :param total_docs: Total number of documents
+    :return: dict of properties describing the corpus for templates
+    """
 
     logger = logging.getLogger(__name__)
-    freq = max(1, number_of_docs // 1000)
 
-    progress = console.progress()
     compressor = DOCS_COMPRESSOR()
+    out_path = f"{out_path}_{start_doc}_{end_doc}" + OUT_EXT
     comp_outpath = out_path + COMP_EXT
+
     with open(out_path, "wb") as outfile:
         with open(comp_outpath, "wb") as comp_outfile:
-            logger.info("Dumping corpus for index [%s] to [%s].", index, out_path)
-            query = {"query": {"match_all": {}}}
-            for n, doc in enumerate(helpers.scan(client, query=query, index=index)):
-                if n >= number_of_docs:
+            logger.info(
+                "Dumping corpus for index [%s] to [%s] for docs %s-%s.",
+                index,
+                out_path,
+                start_doc,
+                end_doc,
+            )
+            query = {
+                "query": {"match_all": {}},
+                "from": start_doc,
+                "size": end_doc - start_doc,
+            }
+
+            batch_size = (end_doc - start_doc) // 5
+            search_after = None
```

Comment:
When will `search_after` not be None?

Reply:
Like everytime the document has been retrieved, search_after gets set to that. This keeps going on until one batch is over. So in the next call for a batch, search_after becomes the (last document retreieved + 1)